### PR TITLE
Fix IDEA-57695

### PR DESCRIPTION
**Pull request by zuevmaxim**
https://youtrack.jetbrains.com/issue/IDEA-57695

If a LINENUMBER instruction for the same line appeared twice, jump counter in `LineEnumerator` had reset to zero. Then `removeLastJump` operation caused other jump removal.

Now labels are used instead of counters for jumps identification.

### CODE REVIEW

File: instrumentation/src/com/intellij/rt/coverage/instrumentation/LineEnumerator.java [hunk 1]
```diff
@@ -313,4 +334,53 @@ public String getDescriptor() {
   private static List<LineEnumeratorFilter> createLineEnumeratorFilters() {
     return KotlinUtils.createLineEnumeratorFilters();
   }
+
+  static class Jump {
```
Comment (akozlova) [inline instrumentation/src/com/intellij/rt/coverage/instrumentation/LineEnumerator.java:338]: I would add equals/hashCode for the case you would compare objects one day

File: instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java [hunk 2]
```diff
@@ -73,221 +57,71 @@ public void visitLabel(Label label) {
...
+    pushIntValue(line);
+    pushIntValue(jumpIndex);
     mv.visitInsn(trueHit ? Opcodes.ICONST_0 : Opcodes.ICONST_1);
     mv.visitMethodInsn(Opcodes.INVOKESTATIC, ProjectData.PROJECT_DATA_OWNER, "touchJump", "(Ljava/lang/Object;IIZ)V", false);
-
-    mv.visitIntInsn(Opcodes.SIPUSH, -1);
-    mv.visitVarInsn(Opcodes.ISTORE, getJumpVariableNumber());
-  }
-
-  private void touchLastJump() {
```
Comment (akozlova) [inline instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: if I remember correctly, this was used to hit false branch for cases like
```
if (true) {
  return xxx;
}
doSomth();
return zzz;
```

is it covered now?

Comment (zuevmaxim) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: It is covered like this
```
if (true) {   // PARTIAL
  return xxx; // FULL
}
doSomth();    // NONE
return zzz;   // NONE
```
or with `false`
```
if (false) {   // PARTIAL
  return xxx; // NONE
}
doSomth();    // FULL
return zzz;   // FULL
```
Or if code is run twice with both `true/false`
```
if (b) {   // FULL
  return xxx; // FULL
}
doSomth();    // FULL
return zzz;   // FULL
```

Comment (akozlova) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: Oh no, I've asked how your code deal with this user sample, not about resulted coverage. Pan was not intended, sorry

Comment (zuevmaxim) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: It should be covered as jumps (`IF_EQ label`) are transformed like this now:
```
IF_EQ true_label
GOTO false_label
true_label:
  touchTrue()
  GOTO label
false_label:
  touchFalse()
  // false code here
```

Comment (zuevmaxim) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: The point of this example is the way of touching false branch when `doSomth` is called(but only if condition is false). Jump to it's label is changed to jump at internal label where touch is done, after that a jump to the original label is done. Therefore false touch is done only if this jump is done.

Comment (akozlova) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: Is it this code:
// a super call to filters may remove this jump, so myLastJump may be null after this line
         super.visitJumpInsn(opcode, trueLabel);
         super.visitJumpInsn(Opcodes.GOTO, falseLabel);
         super.visitLabel(trueLabel);  // true hit will be inserted here
         super.visitJumpInsn(Opcodes.GOTO, label);
         super.visitLabel(falseLabel); // false hit will be inserted here

Comment (zuevmaxim) [reply instrumentation/src/com/intellij/rt/coverage/instrumentation/TouchCounter.java:134]: Yes, it is

**Review by akozlova — APPROVED** [review]

**Comment by akozlova** [comment]
I think, it's merged now and I close the PR (it had some strange state for me)

**Comment by zuevmaxim** [comment]
It has not been merged yet. It waits for [coroutines PR](https://github.com/JetBrains/intellij-coverage/pull/20) to be merged first

**Pull request merged**